src/main/java/com/aws/greengrass/mqttclient/MqttClient.java

```diff
@@ -433,50 +445,58 @@ public synchronized void unsubscribe(UnsubscribeRequest request)
-                final long id = spool.popId();
-                PublishRequest request = spool.getMessageById(id);
-                if (request == null) {
-                    continue;
-                }
-
-                MqttMessage m = new MqttMessage(request.getTopic(), request.getPayload());
-                publishRequests.add(getConnection(false).publish(m, request.getQos(), request.isRetain())
-                                .whenComplete((packetId, throwable) -> {
+        MqttMessage m = new MqttMessage(request.getTopic(), request.getPayload());
+        return getConnection(false).publish(m, request.getQos(), request.isRetain())
+                .whenComplete((packetId, throwable) -> {
                     // packetId is the SDK assigned ID. Ignore this and instead use the spooler ID
                     if (throwable == null) {
                         spool.removeMessageById(id);
                         logger.atDebug().kv("id", id).kv("topic", request.getTopic())
-                            .log("Successfully published message");
+                                .log("Successfully published message");
                     } else {
                         spool.addId(id);
                         logger.atError().log("Failed to publish the message via Spooler", throwable);
                     }
-                }));
+                });
+    }
+
+    /**
+     * Iterate the spooler queue to publish all the spooled message.
+     */
+    protected void publishAllTasks() {
+        while (!Thread.currentThread().isInterrupted()) {
+            try {
+                getConnection(false).connect().get();
+                List<CompletableFuture<?>> publishRequests = new ArrayList<>();
+                while (waitMqttOnline.get()) {
+                    CompletableFuture future = publishSingleTask();
+                    if (future == null) {
+                        continue;
+                    }
+                    publishRequests.add(future);
 
-                if (publishRequests.size() >= maxInFlightPublishes) {
-                    CompletableFuture.anyOf(
-                            publishRequests.toArray(new CompletableFuture[0])).get();
-                    publishRequests = publishRequests.stream().filter(f -> !f.isDone()).collect(Collectors.toList());
+                    if (publishRequests.size() >= maxInFlightPublishes) {
+                        CompletableFuture.anyOf(
+                                publishRequests.toArray(new CompletableFuture[0])).get();
+                        publishRequests = publishRequests.stream().filter(f ->
+                                !f.isDone()).collect(Collectors.toList());
+                    }
                 }
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                if (spoolingFuture.get() != null) {
+                    spoolingFuture.get().cancel(true);
+                }
```

Comment:
Don't do this. The only way the thread is interrupted is if the task was cancelled already. This thread should not be managing itself, that's what the outside class is already doing.